lib/scholar/manifold/tsne.ex

```diff
@@ -0,0 +1,302 @@
+      type: {:or, [:pos_integer, :float]},
+      default: 500,
+      doc: ~S"""
+      The learning rate for t-SNE is usually in the range [10.0, 1000.0].
+      If the learning rate is too high, the data may look like a 'ball' with any point approximately equidistant from its nearest neighbors.
+      If the learning rate is too low, most points may look compressed in a dense cloud with few outliers.
+      If the cost function gets stuck in a bad local minimum increasing the learning rate may help.
+      """
+    ],
+    num_iters: [
+      type: :pos_integer,
+      default: 500,
+      doc: ~S"""
+      Maximum number of iterations for the optimization.
+      """
+    ],
+    seed: [
+      type: :integer,
+      doc: """
+      Determines random number generation for centroid initialization.
+      If the seed is not provided, it is set to `System.system_time()`.
+      """
+    ],
+    init: [
+      type: {:in, [:random, :pca]},
+      default: :pca,
+      doc: ~S"""
+      Initialization of embedding.
+      Possible options are `:random` and `:pca`.
+      """
+    ],
+    metric: [
+      type: {:in, [:euclidean, :squared_euclidean, :cosine, :manhattan, :chebyshev]},
+      default: :squared_euclidean,
+      doc: ~S"""
+      Metric used to compute the distances.
+      """
+    ],
+    exaggeration: [
+      type: {:or, [:float, :pos_integer]},
+      default: 10.0,
+      doc: ~S"""
+      Controls how tight natural clusters in the original space are in the embedded space and
+      how much space will be between them. For larger values, the space between natural clusters will be larger in the embedded space.
+      """
+    ]
+  ]
+
+  @opts_schema NimbleOptions.new!(opts_schema)
+
+  @doc """
+  Fits tSNE for sample inputs `x`.
+
+  ## Options
+
+  #{NimbleOptions.docs(@opts_schema)}
+
+  ## Return Values
+
+  Returns the embedded data `y`.
+
+  ## Examples
+
+      iex> x = Nx.iota({4,5})
+      iex> seed = 42
+      iex> Scholar.Manifold.TSNE.fit(x, num_components: 2, seed: seed)
+      #Nx.Tensor<
+        f32[4][2]
+        [
+          [11.721418380737305, 0.0],
+          [74.18099975585938, 0.0],
+          [-24.942012786865234, 0.0],
+          [-60.960227966308594, 0.0]
+        ]
+      >
+  """
+  deftransform fit(x, opts \\ []) do
+    opts = NimbleOptions.validate!(opts, @opts_schema)
+    seed = Keyword.get_lazy(opts, :seed, &System.system_time/0)
+    fit_n(x, seed, opts)
+  end
+
+  defnp fit_n(x, seed, opts \\ []) do
+    {perplexity, learning_rate, num_iters, num_components, exaggeration, init, metric} =
+      {opts[:perplexity], opts[:learning_rate], opts[:num_iters], opts[:num_components],
+       opts[:exaggeration], opts[:init], opts[:metric]}
+
+    x = to_float(x)
+    {n, _dims} = Nx.shape(x)
+
+    y1 =
+      case init do
+        :random ->
+          key = Nx.Random.key(seed)
+
+          {y, _new_key} =
+            Nx.Random.normal(key, 0.0, 1.0e-4, shape: {n, num_components}, type: Nx.type(x))
```

Comment:
forgot to ask: any reason for `1.0e-4`?

Reply:
(not a blocker for the PR tho)

Reply:
[Simply because I suggest how they initialize it in sklearn](https://github.com/scikit-learn/scikit-learn/blob/364c77e04/sklearn/manifold/_t_sne.py#L994). Btw, you remind me that I forgot to rescale initial embeddings for pca initialization 😅